tools/jit/templates/aten_dispatch.cpp

```diff
@@ -20,19 +20,53 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
 
 namespace {
 
-void pack_list(std::vector<Tensor> & outputs, Tensor v) { outputs.push_back(v); }
-void pack_list(std::vector<Tensor> & outputs, Scalar v) { outputs.push_back(v.toTensor()); }
-void pack_list(std::vector<Tensor> & outputs, const std::vector<Tensor> & t) {
-  outputs.insert(outputs.end(), t.begin(), t.end());
+// a temporary Tensor that does not alter the refcount of impl on
+// acquisition or release, avoids any refcounting in dispatch functions
+struct TensorTemporary : public at::Tensor {
+  explicit TensorTemporary(at::RefCounted * impl)
+  : at::Tensor(static_cast<at::TensorImpl*>(impl), false /* do not retain*/) {}
+  ~TensorTemporary() {
+    detach(); // reset
+  }
+};
+
+// same thing but creates a tensor list, only used rarely (e.g. for cat)
+struct TensorTemporaryList {
```

Comment:
I had to do some mental work to convince myself this was safe, it might be worth writing down: the key is that a `TensorTemporaryList` only exposes a `TensorList`, not its underlying `std::vector<at::Tensor>`. This ArrayRef has the desired semantics: if you get out an `at::Tensor` from it, the refcount is bumped; if you take a reference, it is only guaranteed to stay live as long as the ArrayRef is live, which is exactly the guarantee you need to use this safely.